### PR TITLE
CHAIN-279 make order book level checkpoint respect circular buffer

### PR DESCRIPTION
**Pull request by imykolenko**

### CODE REVIEW

File: sequencer/src/main/kotlin/co/chainring/sequencer/core/Market.kt [hunk 1]
```diff
@@ -476,10 +485,20 @@ data class Market(
...
+                }
                 RemoveOrderResult(levelOrder.wallet, BigInteger.ZERO, notionalPlusFee(levelOrder.quantity, level.price, baseDecimals, quoteDecimals, feeRates.maker))
             } else {
-                sellOrdersByWallet[levelOrder.wallet]?.remove(levelOrder)
+                sellOrdersByWallet[levelOrder.wallet]?.let {
+                    it.remove(levelOrder)
+                    if (it.isEmpty()) {
+                        sellOrdersByWallet.remove(levelOrder.wallet)
+                    }
+                }
```
Comment (imykolenko) [inline sequencer/src/main/kotlin/co/chainring/sequencer/core/Market.kt:501]: Cleaning up empty list when the last order for the wallet is removed. This allows to directly compare initial/restored sequencer state. 

Other approach would be to implement smarter `equals`, but I though maybe cleaning state is better.

**Review by bholzman — APPROVED** [review]
Looks good!

File: sequencer/src/main/kotlin/co/chainring/sequencer/core/Market.kt [hunk 2]
```diff
@@ -867,12 +886,10 @@ data class Market(
             this.maxOfferIx = this@Market.maxOfferIx
             this.bestBid = this@Market.bestBid.toDecimalValue()
             this.bestOffer = this@Market.bestOffer.toDecimalValue()
-            val maxBidIx = (this@Market.bestBid - this@Market.levels[0].price).divideToIntegralValue(this@Market.tickSize).toInt()
-            val minOfferIx = (this@Market.bestOffer - this@Market.levels[0].price).divideToIntegralValue(this@Market.tickSize).toInt()
-            val firstLevelWithData = this.minBidIx.let { if (it == -1) minOfferIx else it }
-            val lastLevelWithData = this.maxOfferIx.let { if (it == -1) maxBidIx else it }
-            (firstLevelWithData..lastLevelWithData).forEach { i ->
-                this.levels.add(this@Market.levels[i].toCheckpoint())
+            this@Market.levels.forEach { level ->
```
[the remark this answers is not in the document]
Comment (imykolenko) [reply sequencer/src/main/kotlin/co/chainring/sequencer/core/Market.kt:889]: you right, that can be optimised.
I will narrow iteration only to levels between `minBidIx` and `maxOfferIx`, but still will only include levels with data.

[the remark this answers is not in the document]
Comment (imykolenko) [reply sequencer/src/main/kotlin/co/chainring/sequencer/core/Market.kt]: done

**Pull request merged**